src/modules/bulletproofs/inner_product_impl.h

```diff
@@ -0,0 +1,848 @@
+    secp256k1_gej r;
+    secp256k1_scalar zero;
+    size_t i;
+
+    if (plen != secp256k1_bulletproof_innerproduct_proof_length(vec_len)) {
+        return 0;
+    }
+
+    if (n_proofs == 0) {
+        return 1;
+    }
+
+    if (!secp256k1_scratch_allocate_frame(scratch, n_proofs * (sizeof(*ecmult_data.randomizer) + sizeof(*ecmult_data.proof)), 2)) {
+        return 0;
+    }
+
+    secp256k1_scalar_clear(&zero);
+    ecmult_data.n_proofs = n_proofs;
+    ecmult_data.g = gens->blinding_gen;
+    ecmult_data.geng = gens->gens;
+    ecmult_data.genh = gens->gens + gens->n / 2;
+    ecmult_data.vec_len = vec_len;
+    ecmult_data.lg_vec_len = secp256k1_floor_lg(2 * vec_len / IP_AB_SCALARS);
+    ecmult_data.shared_g = shared_g;
+    ecmult_data.randomizer = (secp256k1_scalar *)secp256k1_scratch_alloc(scratch, n_proofs * sizeof(*ecmult_data.randomizer));
+    ecmult_data.proof = (secp256k1_bulletproof_innerproduct_vfy_data *)secp256k1_scratch_alloc(scratch, n_proofs * sizeof(*ecmult_data.proof));
+    /* Seed RNG for per-proof randomizers */
+    secp256k1_sha256_initialize(&sha256);
+    for (i = 0; i < n_proofs; i++) {
+        secp256k1_sha256_write(&sha256, proof[i].proof, plen);
+        secp256k1_sha256_write(&sha256, proof[i].commit, 32);
+        secp256k1_scalar_get_b32(commit, &proof[i].p_offs);
+        secp256k1_sha256_write(&sha256, commit, 32);
+    }
+    secp256k1_sha256_finalize(&sha256, commit);
+
+    secp256k1_scalar_clear(&ecmult_data.p_offs);
+    for (i = 0; i < n_proofs; i++) {
+        const unsigned char *serproof = proof[i].proof;
+        unsigned char proof_commit[32];
+        secp256k1_scalar dot;
+        secp256k1_scalar ab[IP_AB_SCALARS];
+        secp256k1_scalar negprod;
+        secp256k1_scalar x;
+        int overflow;
+        size_t j;
+        const size_t n_ab = 2 * vec_len < IP_AB_SCALARS ? 2 * vec_len : IP_AB_SCALARS;
+
+        total_n_points += 2 * ecmult_data.lg_vec_len + proof[i].n_extra_rangeproof_points - !!shared_g; /* -1 for shared G */
+
+        /* Extract dot product, will always be the first 32 bytes */
+        secp256k1_scalar_set_b32(&dot, serproof, &overflow);
+        if (overflow) {
+            secp256k1_scratch_deallocate_frame(scratch);
+            return 0;
+        }
+        /* Commit to dot product */
+        secp256k1_sha256_initialize(&sha256);
+        secp256k1_sha256_write(&sha256, proof[i].commit, 32);
+        secp256k1_sha256_write(&sha256, serproof, 32);
+        secp256k1_sha256_finalize(&sha256, proof_commit);
+        serproof += 32;
+
+        /* Extract a, b */
+        for (j = 0; j < n_ab; j++) {
+            secp256k1_scalar_set_b32(&ab[j], serproof, &overflow);
+            if (overflow) {
+                secp256k1_scratch_deallocate_frame(scratch);
+                return 0;
+            }
+            /* TODO our verifier currently bombs out with zeros because it uses
```

Comment:
I don't understand this.

Reply:
If a_1, a_2, b_1, b_2 are 0, then later in the code where we compute `abinv`, we would try to take a scalar inverse of 0. Fixing this would require changing the whole algorithm, so instead we simply reject such proofs. (They happen with probability ~0 when doing anything zero-knowledge, since in that case every component of the input vectors are uniformly random.)